wxGUI/tplot: add option to show simple regression model line

tmszi opened the pull request:
Add option to show simple regression model line (calculate simple regression model, print calculated formula y = a + b*x, and r-squared parameter). Implements the issue #1197.

neteler commented:
It looks nice. I have locally changed from 4 to 5 digits, to show more details (see screenshot).

Just the computation of `a` I am not yet sure about: in R and LibreOffice I get the different (not perfectly identical) value of ~96 instead of 104.xx - see screenshot:

![image](https://user-images.githubusercontent.com/1295172/102943058-4c9d1900-44b7-11eb-8b17-8afff425be75.png)

tmszi commented:
> Just the computation of `a` I am not yet sure about: in R and LibreOffice I get the different (not perfectly identical) value of ~96 instead of 104.xx 

Main problem why results are different ('g.gui.tplot' vs. R) is that CSV file has different values for x axis than those that enter into drawing plot and calculation simple linear regression model in g.gui.tplot


NC location tempmean raster dataset (climate 2002-2012)

I export x axis data (xdata variable) and  y axis data (ydata variable) [tempmean.zip](https://github.com/OSGeo/grass/files/5734075/tempmean.zip) (I added code which export data, before this [line](https://github.com/OSGeo/grass/commit/2c9a4955897dc571e12ee38c5bba1bc9500f7fc0#diff-fb872d458f8eefd55c65756bf36e0584432c37295748d47e1a34e5bb010b767fR714)) and not used exported data (g.gui.tplot CSV export option) from CSV file:


**g.gui.tplot:**

` g.gui.tplot -l strds=tempmean@climate_2000_2012 coordinates=641266.2256682896,225208.47197099228`

Result:

y = -161.3596 + 0.0002*x
r² = 0.0018

**R:**

```
> x <- c(730120.0, 730151.0, 730180.0, 730211.0, 730241.0, 730272.0, 730302.0, 730333.0, 730364.0, 730394.0, 730425.0, 730455.0, 730486.0, 730517.0, 730545.0, 730576.0, 730606.0, 730637.0, 730667.0, 730698.0, 730729.0, 730759.0, 730790.0, 730820.0, 730851.0, 730882.0, 730910.0, 730941.0, 730971.0, 731002.0, 731032.0, 731063.0, 731094.0, 731124.0, 731155.0, 731185.0, 731216.0, 731247.0, 731275.0, 731306.0, 731336.0, 731367.0, 731397.0, 731428.0, 731459.0, 731489.0, 731520.0, 731550.0, 731581.0, 731612.0, 731641.0, 731672.0, 731702.0, 731733.0, 731763.0, 731794.0, 731825.0, 731855.0, 731886.0, 731916.0, 731947.0, 731978.0, 732006.0, 732037.0, 732067.0, 732098.0, 732128.0, 732159.0, 732190.0, 732220.0, 732251.0, 732281.0, 732312.0, 732343.0, 732371.0, 732402.0, 732432.0, 732463.0, 732493.0, 732524.0, 732555.0, 732585.0, 732616.0, 732646.0, 732677.0, 732708.0, 732736.0, 732767.0, 732797.0, 732828.0, 732858.0, 732889.0, 732920.0, 732950.0, 732981.0, 733011.0, 733042.0, 733073.0, 733102.0, 733133.0, 733163.0, 733194.0, 733224.0, 733255.0, 733286.0, 733316.0, 733347.0, 733377.0, 733408.0, 733439.0, 733467.0, 733498.0, 733528.0, 733559.0, 733589.0, 733620.0, 733651.0, 733681.0, 733712.0, 733742.0, 733773.0, 733804.0, 733832.0, 733863.0, 733893.0, 733924.0, 733954.0, 733985.0, 734016.0, 734046.0, 734077.0, 734107.0, 734138.0, 734169.0, 734197.0, 734228.0, 734258.0, 734289.0, 734319.0, 734350.0, 734381.0, 734411.0, 734442.0, 734472.0, 734503.0, 734534.0, 734563.0, 734594.0, 734624.0, 734655.0, 734685.0, 734716.0, 734747.0, 734777.0, 734808.0, 734838.0)

> y <- c(4.177794986300999, 7.831804487440321, 12.327876620822483, 14.533443450927734, 21.317626105414497, 24.817865159776474, 24.75746578640408, 24.592102898491753, 21.144676208496094, 15.866307152642143, 8.871498107910156, 2.0016182793511286, 4.481419457329644, 8.147710164388021, 8.638227250840929, 16.102625528971355, 19.7551515367296, 24.498782687717014, 24.249119228786892, 25.42009989420573, 20.456763373480904, 14.85820558336046, 13.069485558403862, 9.145535363091362, 5.997615390353733, 7.608275943332249, 10.752396053738064, 17.846107482910156, 19.301422966851128, 25.03023359510634, 26.840180291069878, 25.835588243272568, 22.81259324815538, 16.92228741115994, 9.40331776936849, 4.607234530978733, 2.628127204047309, 5.222822825113933, 11.494742499457466, 14.771919250488281, 19.363984002007378, 23.16556718614366, 25.77496846516927, 25.9027099609375, 21.549483405219185, 15.4322878519694, 13.67812050713433, 5.023538801405165, 3.3117824130588107, 4.819950527615017, 11.154842376708984, 15.677710639105904, 22.571178012424046, 24.21041700575087, 26.152140299479168, 23.878601921929253, 21.74257490370009, 16.718563503689236, 11.871176825629341, 5.748280419243707, 6.019526587592231, 6.646253797743055, 8.527183532714844, 15.018895467122396, 18.07671440972222, 23.78995683458116, 27.271291944715713, 26.213374667697483, 24.0997314453125, 16.979658338758682, 11.277398003472221, 4.635100894504124, 8.352875179714626, 5.861091613769531, 10.294375949435764, 17.124466366238064, 18.45818413628472, 23.314950731065537, 26.111649407280815, 26.370489332411026, 20.750838385687935, 14.720247056749132, 11.195316314697266, 8.52784898546007, 7.373458014594184, 4.040169186062283, 12.215843200683594, 14.930733574761284, 19.572410583496094, 24.14621141221788, 25.323740641276043, 28.032455444335938, 23.270878261990017, 19.105266994900173, 10.133037567138672, 8.716131846110025, 4.7563425699869795, 8.097048865424263, 10.95112058851454, 15.297455257839626, 18.776880900065105, 26.198755900065105, 25.631518893771702, 24.978421529134113, 22.055952284071182, 14.990562862820095, 9.064405229356554, 7.923588222927517, 3.781460656060113, 6.477813720703125, 9.746701982286242, 15.681997934977213, 20.499339633517796, 24.602690802680122, 24.993587070041233, 26.275257534450954, 21.32695092095269, 15.466251373291016, 11.777875688340929, 4.861384497748481, 2.910669114854601, 3.074707455105252, 10.548992156982422, 17.16406716240777, 21.701914469401043, 26.494759453667534, 27.291259765625, 26.44995371500651, 23.960766262478298, 16.822630564371746, 10.167306264241537, 1.278487311469184, 2.892400953504774, 7.768514421251085, 10.443005032009548, 16.962250603569878, 20.494935777452255, 25.888137817382812, 27.797058953179253, 26.151703728569878, 22.5440427992079, 15.112298329671225, 11.798059675428602, 8.461112976074219, 6.9353442721896705, 7.796609666612413, 15.086780124240452, 15.377120971679688, 21.48746066623264, 22.955267164442276, 27.999627855088974, 25.000059339735245, 21.514205932617188, 15.834534962972006, 7.89655261569553, 9.08884260389)

> result <- lm(y~x)

> result

Call:
lm(formula = y ~ x)

Coefficients:
(Intercept)            x  
 -1.614e+02    2.419e-04  

> summary(result)

Call:
lm(formula = y ~ x)

Residuals:
     Min       1Q   Median       3Q      Max 
-14.9343  -7.1822  -0.2654   7.5288  12.1143 

Coefficients:
              Estimate Std. Error t value Pr(>|t|)
(Intercept) -1.614e+02  3.327e+02  -0.485    0.628
x            2.419e-04  4.542e-04   0.533    0.595

Residual standard error: 7.775 on 154 degrees of freedom
Multiple R-squared:  0.001838,	Adjusted R-squared:  -0.004643 
F-statistic: 0.2836 on 1 and 154 DF,  p-value: 0.5951
```

The results are identical.

neteler commented:
> Main problem why results are different ('g.gui.tplot' vs. R) is that CSV file has different values for x axis than those that enter into drawing plot and calculation simple linear regression model in g.gui.tplot

Maybe I am missing something but shouldn't the values for drawing the plot and the linear regression model calculation be identical? Like this it sounds to be a bit confusing...

tmszi commented:
> > Main problem why results are different ('g.gui.tplot' vs. R) is that CSV file has different values for x axis than those that enter into drawing plot and calculation simple linear regression model in g.gui.tplot
> 
> Maybe I am missing something but shouldn't the values for drawing the plot and the linear regression model calculation be identical? Like this it sounds to be a bit confusing...

Data for drawing plot and calculate simple linear regression are identical in g.gui.tplot. But if you try export data from g.gui.tplot as CVS file  data format for x axis is different.

Data which enter for drawing plot and calculating simple regression model:
[tempmean.zip](https://github.com/OSGeo/grass/files/5736563/tempmean.zip)

e.g. first row
`730120.0,4.177794986300999`

Exported data from g.gui.tplot dialog as CSV file:
[data_exported_from_tplot.zip](https://github.com/OSGeo/grass/files/5736577/data_exported_from_tplot.zip)

e.g. first row
`2000-01-01 00:00:00,4.177794986300999`

veroandreo commented:
It's a matter of date formatting, I believe. When exporting, it seems that date is converted to this format that represents number of days from Jan 1st, 1979 or so. This is convenient to estimate the linear regression between the variable of interest and time.

My 0.2 cents

neteler commented:
I wrote a small test case (for the usual North Carolina sample dataset from [here](https://grass.osgeo.org/download/data/#NorthCarolinaDataset)):

```
export GRASS_OVERWRITE=1

# just a small region in NC
g.region vector=urbanarea -p res=5000 -a

for map in $(seq 1 100) ; do
 # generate synthetic maps with linearily increasing values
 r.mapcalc -s expression="test_${map} = $map"
 echo test_${map} >> map_list.txt
done

# validate
r.info -r test_1
#min=1
#max=1
r.info -r test_80
#min=80
#max=80
r.info -r test_100
#min=100
#max=100


t.create type=strds temporaltype=absolute \
         output=test \
         title="Linear increasing values" \
         description="Test dataset with linear increasing values"

t.register -i type=raster input=test \
           file=map_list.txt start="2020-01-01" increment="1 days"

# check general information of the daily strds
t.info type=strds input=test

# get center of region coordinates for query
eval $(g.region -cg)

g.gui.tplot -l strds=test coordinates=$center_easting,$center_northing
```
![image](https://user-images.githubusercontent.com/1295172/103042115-cf39dd00-4578-11eb-984a-c9ce4a16a77a.png)

neteler commented:
So, the intercept is at 1970-01-01 (not sure why I am off by one day but hey):

```
date -d '2020-01-01 - 18261 days' +"%Y-%m-%d"
1970-01-02
```

Maybe worth an addition then in the manual that it is there reference date...

tmszi commented:
> Maybe worth an addition then in the manual that it is there reference date...
> Or show it as a date in the equation, @tmszi ?

I mentioned it in the manual page, commit [fa8b8e](https://github.com/OSGeo/grass/pull/1207/commits/fa8b8e41c66829ff90ae5a50aba2f9aae9dc1ed7). Is it OK like this?

neteler commented on gui/wxpython/tplot/g.gui.tplot.html:
```diff
@@ -18,7 +18,13 @@ <h2>DESCRIPTION</h2>
+    if you want to use for calculating simple regression model in the
+    <a href="https://www.r-project.org/">R environment<a>,
+    <a href="https://www.libreoffice.org/">LibreOffice</a>
+    etc., you get different calculated formula <pre>y = a + b*x</pre> because they use
+    a reference date other than the UNIX Epoch time (00:00:00 UTC on 1 January 1970)).
```
```suggestion
    etc., you will obtain a different calculated formula <pre>y = a + b*x</pre> because these
    software packages use a reference date other than the UNIX Epoch time (00:00:00 UTC on 1 January 1970)).
```

neteler reviewed (approved):
Thanks for the implementation, @tmszi!
